src/memory/pageTable.ts

```diff
@@ -0,0 +1,135 @@
+import assert from 'assert';
+import { MemoryError, MemoryErrorType } from './memoryError';
+
+export class PageTable {
+  public static readonly NUM_OF_ENTRIES: number = 512;
+  public static readonly ENTRY_SIZE: number = 8;
+  private static readonly EMPTY_FREE_LIST: number = -1;
+
+  private readonly memory: DataView;
+  private freeList: number;
+
+  constructor() {
+    const data = new ArrayBuffer(
+      PageTable.NUM_OF_ENTRIES * PageTable.ENTRY_SIZE
+    );
+    this.memory = new DataView(data);
+
+    // The free list is made of entries, where each free entry carries the offset of the next free entry
+    this.freeList = 0;
+    let i;
+    for (i = this.freeList; i < PageTable.NUM_OF_ENTRIES - 1; i++) {
+      this.memory.setFloat64(i * PageTable.ENTRY_SIZE, i + 1);
+    }
+    this.memory.setFloat64(i, PageTable.EMPTY_FREE_LIST);
+  }
+
+  /**
+   * Gets the data from the entry at the offset.
+   * Note: This should be the offset to an entry (9 bits),
+   * not an offset to the within the entry (12 bits)
+   */
+  public get(offset: number): number {
+    if (!this.isValidOffset(offset)) {
+      throw new MemoryError(MemoryErrorType.INVALID_OFFSET, offset);
+    }
+
+    return this.memory.getFloat64(offset * PageTable.ENTRY_SIZE);
+  }
+
+  /**
+   * Sets the first free entry in the free list to contain data.
+   */
+  public setFreeEntry(data: number): number {
+    const offset = this.freeList;
+
+    if (!this.isValidOffset(offset)) {
+      throw new MemoryError(MemoryErrorType.INVALID_OFFSET, offset);
+    }
+
+    this.freeList = this.get(offset);
+    this.memory.setFloat64(offset * PageTable.ENTRY_SIZE, data);
+    return offset;
+  }
+
+  /**
+   * Sets a free entry to contain data.
+   */
+  public setFreeEntryAt(offset: number, data: number): number {
+    if (!this.isValidOffset(offset)) {
+      throw new MemoryError(MemoryErrorType.INVALID_OFFSET, offset);
+    }
+
+    assert(this.isFree(offset), 'Offset must be free.');
```

Comment:
I wrote this as assert at first, because I think the burden of ensuring this method is used properly lies on the developer. So it's more for development checks.

Checking if something is free requires searching through the entire free list, because if we want to model memory usage closely then we can't store a flag to indicate that the node is free anywhere else.

This means that all the asserts are slow. Also I found out Javascript cannot really disable asserts, but we implement the disabling ourselves. So, should we have this check or not?

Reply:
> This means that all the asserts are slow (but also not slow if we call the method properly on an actual free entry, cos it'll be like O(1)).

If that's the case, I think it's okay to leave it in. It only writes to console if the assertion fails, right? Meaning that if everything is right, having it in the code should result in no change in noticeable behaviour.

Reply:
> This means that all the asserts are slow (but also not slow if we call the method properly on an actual free entry, cos it'll be like O(1)).

If that's the case, I think it's okay to leave it in. It only writes to console if the assertion fails, right? Meaning that if everything is right, having it in the code should result in no change in noticeable behaviour.

Reply:
Yes